File: jvm/src/test/java/com/nawforce/apexparser/SyntaxErrorCounter.java
package com.nawforce.apexparser;

import org.antlr.v4.runtime.*;

import java.util.AbstractMap;
import java.util.Map;

public class SyntaxErrorCounter extends BaseErrorListener {
    private int numErrors = 0;

    @Override
    public void syntaxError(
            Recognizer<?, ?> recognizer,
            Object offendingSymbol,
            int line,
            int charPositionInLine,
            String msg,
            RecognitionException e) {
        this.numErrors += 1;
    }

    public int getNumErrors() {
        return this.numErrors;
    }

    public static Map.Entry<ApexParser, SyntaxErrorCounter> createParser(String input) {
        ApexLexer lexer = new ApexLexer(new CaseInsensitiveInputStream(CharStreams.fromString(input)));
        CommonTokenStream tokens = new CommonTokenStream(lexer);
        ApexParser parser = new ApexParser(tokens);

        parser.removeErrorListeners();
        SyntaxErrorCounter errorCounter = new SyntaxErrorCounter();
        parser.addErrorListener(errorCounter);

        return new AbstractMap.SimpleEntry<ApexParser, SyntaxErrorCounter>(parser, errorCounter);
    }
}

File: npm/src/__tests__/SyntaxErrorCounter.ts
import { ANTLRErrorListener, CommonTokenStream, RecognitionException, Recognizer, Token } from "antlr4ts";
import { ApexLexer } from "../ApexLexer";
import { ApexParser } from "../ApexParser";
import { CaseInsensitiveInputStream } from "../CaseInsensitiveInputStream";

export class SyntaxErrorCounter implements ANTLRErrorListener<Token> {
    numErrors = 0

    syntaxError(recognizer: Recognizer<Token, any>,
        offendingSymbol: Token, line: number, charPositionInLine: number, msg: string,
        e: RecognitionException | undefined): any {
        this.numErrors += 1;
    }

    getNumErrors(): number {
        return this.numErrors;
    }
}

export function createParser(userData: string, input: string): [ApexParser, SyntaxErrorCounter] {
    const lexer = new ApexLexer(new CaseInsensitiveInputStream(userData, input))
    const tokens = new CommonTokenStream(lexer);
    const parser = new ApexParser(tokens)

    parser.removeErrorListeners()
    const errorCounter = new SyntaxErrorCounter();
    parser.addErrorListener(errorCounter);

    return [parser, errorCounter];
}
